server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java

```diff
@@ -52,97 +68,305 @@
+    private final TransportVerifyShardBeforeCloseAction transportVerifyShardBeforeCloseAction;
     private final ActiveShardsObserver activeShardsObserver;
 
     @Inject
     public MetaDataIndexStateService(ClusterService clusterService, AllocationService allocationService,
                                      MetaDataIndexUpgradeService metaDataIndexUpgradeService,
-                                     IndicesService indicesService, ThreadPool threadPool) {
+                                     IndicesService indicesService, ThreadPool threadPool,
+                                     TransportVerifyShardBeforeCloseAction transportVerifyShardBeforeCloseAction) {
         this.indicesService = indicesService;
         this.clusterService = clusterService;
         this.allocationService = allocationService;
+        this.threadPool = threadPool;
+        this.transportVerifyShardBeforeCloseAction = transportVerifyShardBeforeCloseAction;
         this.metaDataIndexUpgradeService = metaDataIndexUpgradeService;
         this.activeShardsObserver = new ActiveShardsObserver(clusterService, threadPool);
     }
 
-    public void closeIndices(final CloseIndexClusterStateUpdateRequest request, final ActionListener<ClusterStateUpdateResponse> listener) {
+    /**
+     * Closes one or more indices.
+     *
+     * Closing indices is a 3 steps process: it first adds a write block to every indices to close, then waits for the operations on shards
+     * to be terminated and finally closes the indices by moving their state to CLOSE.
+     */
+    public void closeIndices(final CloseIndexClusterStateUpdateRequest request, final ActionListener<AcknowledgedResponse> listener) {
         if (request.indices() == null || request.indices().length == 0) {
             throw new IllegalArgumentException("Index name is required");
         }
+        initiateClosing(request.indices(),  request.masterNodeTimeout(), request.ackTimeout(), listener);
+    }
 
-        final String indicesAsString = Arrays.toString(request.indices());
-        clusterService.submitStateUpdateTask("close-indices " + indicesAsString,
-                new AckedClusterStateUpdateTask<ClusterStateUpdateResponse>(Priority.URGENT, request, listener) {
-            @Override
-            protected ClusterStateUpdateResponse newResponse(boolean acknowledged) {
-                return new ClusterStateUpdateResponse(acknowledged);
-            }
+    /**
+     * Step 1 - Start closing indices by adding a write block
+     *
+     * This step builds the list of indices to close (the ones explicitly requested that are not in CLOSE state) and adds the index block
+     * {@link #INDEX_CLOSED_BLOCK} to every index to close in the cluster state. After the cluster state is published, the shards should
+     * start to reject writing operations and we can proceed with step 2.
+     */
+    private void initiateClosing(final Index[] concreteIndices, @Nullable final TimeValue masterTimeout,
```

Comment:
maybe give these methods more concrete names? e.g. addIndexClosedBlocks, waitForClosedBlocksApplied, and closeRoutingTable?